webpack.config.js

```diff
@@ -233,6 +233,11 @@ module.exports = (argvEnv, argv) => {
         if (target.StandaloneGeckoView) firefoxVariant = 'GeckoView'
         if (target.WKWebview) buildTarget = 'WKWebview'
         if (target.E2E) buildTarget = 'E2E'
+        config.plugins.push(
+            new webpack.DefinePlugin({
+                'webpackEnv.shadowRootMode': JSON.stringify(target.E2E ? 'open' : 'closed'),
```

Comment:
Set a env variable `webpackEnv.shadowRootMode ` which controls how shadow roots were rendered. Puppeteer required an `open` mode for accessing elements inside the shadow root.

Reply:
`process.env.NODE_ENV === 'test'`?

Reply:
`NODE_ENV` is not enough. We need an E2E version build for running E2E testing cases.

Reply:
Now you can set default of `renderInShadowRootSettings` to false in the E2E environment instead of add a new magic variable